keps/sig-network/1669-proxy-terminating-endpoints/README.md

```diff
@@ -0,0 +1,378 @@
+
+## Proposal
+
+This KEP proposes that if all endpoints for a given Service scoped to its traffic policy are terminating (i.e. pod.DeletionTimestamp != nil), then all traffic should be sent to
+terminating Pods that are still Ready. Note that the EndpointSlice API introduced a new condition called "Serving" which is semantically equivalent to "Ready" except that the Ready condition
+must always be "False" for terminating pods for compatibility reasons. For consumers of the EndpointSLice API that want to route traffic strictly based on a Pod's readiness ignoring
+it's terminating state, they should be reading the Serving condition going forward. Below are some examples to help illustrate the proposed behavior:
+
+### Example: only some endpoints terminating when traffic policy is "Cluster"
+
+When the traffic policy is "Cluster" and some endpoints are terminating, all traffic should be routed to the ready endpoints that are not terminating..
+
+### Example: only some endpoints terminating on a node when traffic policy is "Local"
+
+When the traffic policy is "Local" and some endpoints are terminating within a single node, traffic should be routed to ready endpoints on that node that are not terminating.
+
+### Example: all endpoints terminating and traffic policy is "Cluster"
+
+When the traffic policy is "Cluster" and all endpoints are terminating, then traffic should be routed to any terminating endpoint that is ready.
+
+### Example: all endpoints terminating on a node when traffic policy is "Local"
+
+When the traffic policy is "Local" and all endpoints are terminating within a single node, then traffic should be routed to any terminating endpoint that is ready on that node.
+
+
+### Handling terminating endpoints that are not ready.
+
+It is worth noting that traffic should not be routed to terminating pods if their readiness probe is failing, even if it is the only endpoints remaining. This is to give workloads
+the flexibility/control to opt out of this behavior by either exiting immediately or failing the readiness probe when receiving SIGTERM from kubelet. This would also be counter-intuitive
+to the current understanding of readiness probes.
+
+
+### User Stories (optional)
+
+#### Story 1
+
+As a user I would like to do a rolling update of a Deployment fronted by a Service Type=LoadBalancer with ExternalTrafficPolicy=Local.
+If a node that has only 1 pod of said deployment goes into the `Terminating` state, all traffic to that node is dropped until either a new pod
+comes up or my cloud provider removes the node from the loadbalancer's node pool. Ideally the terminating pod should gracefully handle traffic to this node
+until either one of the conditions are satisfied.
+
+### Risks and Mitigations
+
+There are scalability implications to tracking termination state in EndpointSlice. For now we are assuming that the performance trade-offs are worthwhile but
+future testing may change this decision. See KEP 1672 for more details.
```

Comment:
this was brought up during last sig-net meeting, maybe is worth to merge API and implementation in the same kep https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/1672-tracking-terminating-endpoints 
but I really don't know the consequences and the feasibility

Reply:
I originally split them off because there's going to be many ways to consume the new terminating condition outside of the way kube-proxy will use it. For example, ingress controllers may want to implement their own termination logic.

However, I do agree with pacing those two KEPs together since there's a dependency here to KEP-1672